src/core/dataset.jl

```diff
@@ -0,0 +1,148 @@
+abstract type AbstractDataset end
```

Comment:
My main comment is that these types aren't quite polymorphic. They are not going to return the same types. For example, `get_values`. It is returning a `DataFrame` currently. The HDF5Dataset can't do that. Here is a call site:
```
get_values(container[key])[index, :] = df_row
```
You could invert the logic to this:
```
set_values!(container, key, index, df_row)
```
The same function and inputs would work for both both types. The HDF5 could would transform the dataframe row as necessary. The dataframe/dataset would not need to be returned to the caller.

I suggest documenting all of the external interfaces right here. By "external" I mean external to the file where callers don't care whether they have dataframes or hdf5 datasets.

Reply:
fixed in part via 927d781e96adf851a8ed98243142c4a4f2e64289, still need to implement the HDF version of this